scripts/imgtool/keys/ecdsa.py

```diff
@@ -191,3 +209,84 @@ def sign(self, payload):
             return sig
         else:
             return sig
+
+
+class ECDSA384P1Public(ECDSAPublicKey):
+    """
+    Wrapper around an ECDSA (p384) public key.
+    """
+    def __init__(self, key):
+        super().__init__(key)
+        self.key = key
+
+    def shortname(self):
+        return "ecdsap384"
+
+    def sig_type(self):
+        return "ECDSA384_SHA384"
+
+    def sig_tlv(self):
+        return "ECDSASIG"
+
+    def sig_len(self):
+        # Early versions of MCUboot (< v1.5.0) required ECDSA
+        # signatures to be padded to a fixed length.  Because the DER
+        # encoding is done with signed integers, the size of the
+        # signature will vary depending on whether the high bit is set
+        # in each value.  This padding was done in a
+        # not-easily-reversible way (by just adding zeros).
+        #
+        # The signing code no longer requires this padding, and newer
+        # versions of MCUboot don't require it.  But, continue to
+        # return the total length so that the padding can be done if
+        # requested.
+        return 103
+
+    def verify(self, signature, payload):
+        # strip possible paddings added during sign
+        signature = signature[:signature[1] + 2]
+        k = self.key
+        sha = hashlib.sha256()
+        sha.update(payload)
+        digest = sha.digest()
+        if isinstance(self.key, ec.EllipticCurvePrivateKey):
+            k = self.key.public_key()
+        return k.verify(signature=signature, data=digest,
+                        signature_algorithm=ec.ECDSA(SHA384()))
+
+
+class ECDSA384P1(ECDSA384P1Public, ECDSAPrivateKey):
+    """
+    Wrapper around an ECDSA (p384) private key.
+    """
+
+    def __init__(self, key):
+        """key should be an instance of EllipticCurvePrivateKey"""
+        super().__init__(key)
+        self.key = key
+        self.pad_sig = False
+
+    @staticmethod
+    def generate():
+        pk = ec.generate_private_key(
+                ec.SECP384R1(),
+                backend=default_backend())
+        return ECDSA384P1(pk)
+
+    def raw_sign(self, payload):
+        """Return the actual signature"""
+        return self.key.sign(
+                data=payload,
+                signature_algorithm=ec.ECDSA(SHA384()))
```

Comment:
This one is big question. I would be interested, what others think about it? @d3zd3z
AFAICT it is a common/recommended (somewhere required) practice to use a specific ECDSA curve with a hash algorithm of similar level (bits) of security. If P384_SHA256 is acceptable for us, that would be the best interms of code size.

Reply:
Changing the digest will affect other tools that expect an image to use SHA-256, so it should be out of scope of this PR. If you think it needs discussion, please open an issue. SHA-256 is not broken as of today, so until someone finds a pre-image collision, or any related attack we can just stay as is.

Reply:
It's not replacing the SHA-256 image digest, that part is unmodified (see line 283-286). When the image digest is signed, it goes through SHA-384 as well (double hashing). As I read double hashing also happened in the past of MCUboot, when the ECDSA API of the used crypto backend wanted to do the hashing itself (could not be skipped).

If P384_SHA256 is acceptable, the double hashing can be easily avoided. Using SHA-384 for calculating the image digest and replacing SHA-256 seems to have too much impact. The third solution is current one.. double hashing.

Reply:
> SHA-256 is not broken as of today, so until someone finds a pre-image collision, or any related attack we can just stay as is.

But, neither has P-256 been broken. But, increasing the size of the curve without increasing the size of the hash is kind of pointless. There are two general attacks that increasing sizes helps with. One is an increase in raw computation performance, and especially something like quantum computing. The other are vulnerabilities found in the underlying operations that reduce the amount of work needed to attack the operation.

Reply:
>  but then this PR still is P-384(SHA-384(SHA-256(image))). Doesn't that make the SHA-384 just a glorified bit expander?

@utzig: AFAIK the goal is to do the hashing twice: 
- once SHA-256(image) when validating the hash TLV
- 2nd time during the signature verification do P-384(SHA-384(image)). This is going to be done by [psa_verify_message ](https://github.com/Mbed-TLS/mbedtls/blob/development/include/psa/crypto.h#L2939)

Reply:
That would make sense, but at the moment it's not what is being done. And it would be interesting to also have the validation code added to the bootloader, so it's possible to see it working.

Reply:
I've removed the pre hashing of SHA256. Now the image is being hashed as Tamas mentioned. As for the verification code the idea is to if p384 is requested then the hashing should switch to sha384 everywhere.

Reply:
I've updated the imgtool to use sha384 instead of sha256 everywhere if a ECDSAP384 key is being used. There is one thing that needs to be discussed however, that is the TLV. I renamed the sha256 TLV to IMAGEHASH in the imgtool for now but I'd like your input regarding this question if this is acceptable or I should create a dedicated one?
Or better yet, should I just remove the P384 related changes from the PR as it's really getting out of scope and reintroduce them in a separate PR alongside the bootutil changes once the dependent backend is ready?

Reply:
> I guess your argument is "given a random 256-bit string, signing it with with P-384 wouldn't be better than using P-256", which makes sense, but then this PR still is P-384(SHA-384(SHA-256(image))). Doesn't that make the SHA-384 just a glorified bit expander?

Agreed, computing a SHA-384 over a SHA-256 is fairly pointless. You might as well, just use the SHA-256 result with some padding. You aren't gaining any security by adding the extra hash.

I'd have to think about it, but the reduction in work a quantum computer can make over an EC curve will be based on the size of the input, so with a 256-bit input, a quantum computer built to break P-384 would only need the work of P-256, given the smaller input. It isn't likely to be that straightforward, though.

Any other attack against the EC is going to depend on the input size, so I don't understand the point in using P-384 unless we are taking a SHA-384 over the entire image.

BTW, there is no reason that the "image hash" tag needs to be related to how the signature is done. It really is just a shortcut to be able to reject an invalid image without having to verify the curve signature. We're kind of unusual in even storing it, as it is easily computed, and not generally included in signatures. If we want to use P-384, why don't we just compute the SHA384 over the whole image, like we're supposed to?

Reply:
I would also be up for the idea of just removing the hash entirely from the TLV. There isn't really a very good reason for it to be there in the first place. The only thing it really gives us is the ability to reject an invalid image before we have to compute the curve signature. This, for example, could speed up a boot if there is an invalid upgrade image by avoiding having to verify the signature of the upgrade before the primary image could be booted.

Reply:
@d3zd3z @utzig Do you think Roland should modify/split this PR or I can merge it?